tests/importer/importer.cpp

```diff
@@ -0,0 +1,762 @@
+    EXPECT_EQ(fileContents("importer/diamond_point.cellml"), printer->printModel(point));
+}
+
+TEST(Importer, multipleModelResolution)
+{
+    // This test is intended to show how the Importer class can hold multitple imported models in its library, and that
+    // these can be used to resolve the imports of more than one importing model.
+    // The example give has three models A, B, C, each importing the same four components, a, b, c, d from a
+    // fourth model D. We expect the that model D is parsed and instantiated just once, rather than 12 times.
+
+    auto parser = libcellml::Parser::create();
+    auto modelA = parser->parseModel(fileContents("importer/generic.cellml"));
+    auto modelB = parser->parseModel(fileContents("importer/generic.cellml"));
+    auto modelC = parser->parseModel(fileContents("importer/generic.cellml"));
+
+    // ModelD will be imorted into the library as it's included as a dependency in the models here.
+    // Passing in one of the models for resolution will load modelD into the library.
+    auto importer = libcellml::Importer::create();
+    importer->resolveImports(modelA, resourcePath("importer/"));
+
+    EXPECT_EQ(size_t(1), importer->libraryCount());
+
+    // Now resolve the other models and expect the library contents to be unchanged.
+    importer->resolveImports(modelB, resourcePath("importer/"));
+    importer->resolveImports(modelC, resourcePath("importer/"));
+
+    EXPECT_EQ(size_t(1), importer->libraryCount());
+}
+
+TEST(Importer, addModelToLibrary)
+{
+    // This test shows how a model instance can be manually added to the import library by the user,
+    // and will be used to resolve imports, rather than from an external file/URL.
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("importer/generic_no_source.cellml"));
+    auto importer = libcellml::Importer::create();
+
+    auto sourceModel = libcellml::Model::create("source");
+    sourceModel->addComponent(libcellml::Component::create("a"));
+    sourceModel->addComponent(libcellml::Component::create("b"));
+    sourceModel->addComponent(libcellml::Component::create("c"));
+    sourceModel->addComponent(libcellml::Component::create("d"));
+
+    // Add a model manually to the library, including the URL that it will replace in future imports.
+    EXPECT_TRUE(importer->addModel(sourceModel, "i_dont_exist.cellml"));
+
+    // Can't add to the same URL key more than once.
+    EXPECT_FALSE(importer->addModel(sourceModel, "i_dont_exist.cellml"));
+
+    importer->resolveImports(model, "");
+
+    EXPECT_EQ(size_t(0), importer->issueCount());
+    EXPECT_FALSE(model->hasUnresolvedImports());
+}
+
+TEST(Importer, replaceModel)
+{
+    // This test shows how a model instance can be manually added to the import library by the user,
+    // and will be used to resolve imports, rather than from an external file/URL.
```

Comment:
Copied/pasted from the `addModelToLibrary` test above. Needs to be updated accordingly.